# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook **(and nothing else)** named as follows: **StudentName1_snumber_StudentName2_snumber.ipynb**
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Brightspace.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ Use the Brightspace discussion board or email the student assistants for questions on how to complete the exercises.
+ If you find mistakes/have suggestions/would like to complain about the assigment material itself, please email me [Lisa] at `l.tostrams@science.ru.nl`

## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. Check out the script and doc from the Toolbox folder and check if you understand how the association rules are computed. 



In [2]:
import Toolbox.apriorimining as a


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `apriorimining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [4]:
a.generate_association_rules()

Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data/courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.33
set= { 6, 7 },  sup= 83.33
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
--------------------------------------------------------


_We're only looking at rules with 100% confidence, so all we can say is what everyone took - it is Physics._

  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? The script can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [5]:
a.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1

_All the high-confidence association rules are simply for popular movies. The first 20 or so simply spell out "everyone has watched Star Wars, if they even watch movies at all". It's worth noting that this doesn't mean that everyone has watched Star Wars - in fact, only 61% people have. What's more, many association rules involve only the Star Wars movies - perhaps this means that watching one movie in a series make sone more likely to watch the others, but this would require further investigation._

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

_The most-watched movies are Star Wars, Contact and Fargo. However, even they haven't been watched by that many people - this is why there are few rules with many items: it is likely that percentage-wise, there are few people who have watched given movies A, B, and C, given how probabilities work (with probabilities being multiplied for the conjunction of several independent events). Of course, watching movies is not independent events - that's why we're doing data mining at all - but the dependency that can be found in movie series is often lost above 3 anyway, since a trilogy is the  most common *logy._

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

_It is. Using the movie example: if only 2 people in our dataset of 1000 have watched an obscure, rare movie, but they have similar taste - and both watched another, similarly obscure movie, then the itemsets of the two movies will have very low support (0.2%) but 100% confidence._

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

In [6]:
0.3/(0.45 * 0.8)

0.8333333333333333

*1. Confidence measures how often items in b appear in trasactions that contain a, therefore the confidence is 0.3/0.45 = 67%. The rule is interesting because it exceeds the confidence threshold*

*2. The interest measure is 0.3/(0.45 * 0.8) (<1, therefore  the itesms are Negatively correlated)*

*3. High confidence rules may not be interesting*



#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

In [8]:
0.6/(0.9)

0.6666666666666666

*Answers:*
1. First we get the confidence values as follows: 

    $c_1 = \frac{s(a\cup b)}{s(a)}$
    $c_2 = \frac{s(a\cup b \cup c)}{s(a)}$
    $c_3 = \frac{s(a\cup b\cup c)}{s(a \cup c )}$
 
Then, if we consider the set properties to hold:

$$s(a) \geq s(a \cup b) \geq s(a \cup b \cup c)$$, 

Then: $c_1 \geq c_2;  c_3 \geq c_2$, therefore $c_2$ has the lowest confidence. 

2. Yes, it is possible as the confidence values depends on the support of the items.  For example:

$$s(a,b) = 0.5, s(a) = 0.8$$
$$s(b,c) = 0.6, s(b) = 0.7$$
$$s(a,c) = 0.2, s(c) = 0.7$$
Then, let *minconf* = 50%, therefore:
$$c(a \to b) = 62.5\% > minconf$$
$$c(b \to c) = 85.7\% > minconf$$
$$c(a \to c) = 28.5\% < minconf$$


In [11]:
print(0.6/0.7)
print(0.2/0.7)

0.8571428571428572
0.28571428571428575



#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



In [23]:
a = 105; b = 87; c=40; d=62;
OR = (a*d) / (b*c)
print(f'OR table1 :{OR}')

a = 2; b = 9; c=5; d=20;
OR = (a*d) / (b*c)
print(f'OR table2 :{OR}')

a = 103; b = 78; c=35; d=42;
OR = (a*d) / (b*c)
print(f'OR table2 :{OR}')

OR table1 :1.8706896551724137
OR table2 :0.8888888888888888
OR table2 :1.5846153846153845


In [26]:
import math
phi = ((105/294) - (145/294)*(192/294))/(math.sqrt((149/294)*(102/294)*(145/294)*(192/294)))
print(phi)           
t =36
phi = ((2/t) - (7/t)*(11/t))/(math.sqrt((29/t)*(25/t)*(7/t)*(11/t)))
print(phi)         
t =181+77
phi = ((103/t) - (138/t)*(182/t))/(math.sqrt((120/t)*(77/t)*(138/t)*(181/t)))
print(phi)    

0.14730408574297155
-0.0211619477845243
0.0959715939205995


In [27]:
t =294
lift = (105/t)/((145/t)*(192/t))
print(lift)           

t =36
lift = (2/t)/((7/t)*(11/t))
print(lift)    

t =181+77
lift = (103/t)/((138/t)*(181/t))
print(lift)    

1.1088362068965518
0.9350649350649349
1.0638962286812395


*Answers:*

1. odds ratio for table 2 = 1.8706, and for table 3 are 0.8888 and 1.5846
    
2.  The $\phi$ coefficient for both tables is:

 - For table 2, $\phi = 0.1473 $
 - For table 3, $\phi = -0.02116$ and $\phi = 0.09597$
 
3. The interest for both tables is :

 - For table 2,  $ I = 1.10883$
 - For table 3,  $ I = 0.93506$ and $I=1.0638$